ansi/halfblocks/encoder.go

```diff
@@ -0,0 +1,626 @@
+package halfblocks
+
+import (
+	"fmt"
+	"image"
+	"image/color"
+	_ "image/gif"
+	_ "image/jpeg"
+	_ "image/png"
+	"math"
+	"strings"
+)
+
+// Block represents different Unicode block characters
+type Block struct {
+	Char        rune
+	Coverage    []bool // Which parts of the block are filled (true = filled)
+	CoverageMap string // Visual representation of coverage for debugging
+}
+
+// EncoderOptions contains all configurable settings
+type EncoderOptions struct {
+	ColorMode    int     // 0=none, 1=8colors, 2=256colors, 3=truecolor
+	Width        int     // Output width
+	Height       int     // Output height (0 for auto)
+	Threshold    uint8   // Threshold for considering a pixel as set (0-255)
+	DitherLevel  float64 // Dithering amount (0.0-1.0)
+	UseFgBgOnly  bool    // Use only foreground/background colors (no block symbols)
+	InvertColors bool    // Invert colors
+	ScaleMode    string  // fit, stretch, center
```

Comment:
Let's define a type for this one, maybe use `draw.Scaler` https://pkg.go.dev/golang.org/x/image/draw#Scaler